src/main/scala/firrtl/transforms/ConstantPropagation.scala

```diff
@@ -353,6 +353,31 @@ class ConstantPropagation extends Transform with ResolvedAnnotationPaths {
     // AsyncReset registers don't have reset turned into a mux so we must be careful
     val asyncResetRegs = mutable.HashSet.empty[String]
 
+    // A utility class that is somewhat like an Option but with two variants containing Nothing.
+    // UnboundConstant means that a node does not yet have a source of the two allowable types
+    // for register constant propagation (register or literal). BoundConstant means that it has
+    // exactly the one allowable source of that type. NonConstant indicates multiple distinct sources.
+    sealed abstract class ConstPropBinding[+T] {
+      def resolve[V >: T](that: ConstPropBinding[V]): ConstPropBinding[V] = (this, that) match {
+        case (x, y) if (x == y) => x
+        case (x, UnboundConstant) => x
+        case (UnboundConstant, y) => y
+        case _ => NonConstant
+      }
+    }
+    case class BoundConstant[T](value: T) extends ConstPropBinding[T]
+    case object NonConstant extends ConstPropBinding[Nothing]
+    case object UnboundConstant extends ConstPropBinding[Nothing]
```

Comment:
It doesn't matter that much, but maybe these should be hoisted out of this function, perhaps into the companion object (private of course)?